sonic_platform_base/sonic_xcvr/mem_maps/public/cmis.py

```diff
@@ -0,0 +1,48 @@
+"""
+    cmis.py
+
+    Implementation of XcvrMemMap for CMIS Rev 4.0
+"""
+
+from ..xcvr_mem_map import XcvrMemMap
+from ...fields.xcvr_field import (
+    CodeRegField,
+    HexRegField,
+    NumberRegField,
+    RegGroupField,
+    StringRegField,
+)
+from ...fields import consts
+
+PAGE_SIZE = 128
+
+def get_addr(page, offset):
```

Comment:
This seems platform-specific (or at least to be making assumptions about how the data is represented). Is this tied to the Optoe implementation? If so, should it be in the optoe code?

Reply:
There is an assumption of a linear address space, yes. This is provided by Optoe, but the assumption is not tied to it; platforms using other drivers/tools can implement SfpBase.read_eeprom() and SfpBase.write_eeprom() in such a way that this assumption of linearity still works.

Reply:
The current read_eeprom/write_eeprom APIs implicitly assume a linear address space by only accepting a single offset argument, and I chose to do the computation of this single offset in the XcvrMemMap for each XcvrField to accommodate the APIs. This can be dealt with in the platform-specific code if needed, but thinking it over I do agree that it probably makes more sense design wise to only embed page, offset, etc. (whatever is relevant for identifying an address in a spec) in each XcvrField and leave the logic for using those details to access the EEPROM at the SfpOptoeBase abstraction layer.

One idea I have is to define a XcvrAddr absraction, for which there will be a CmisAddr, Sff8436Addr, Sff8472Addr, etc. that derive from it. Each subtype will have relevant attributes for computing an actual address (page, offset, etc.). Then, we replace the read_eeprom/write_eeprom APIs with ones that accept a XcvrAddr instead of an integer offset. SfpOptoeBase would then implement them and handle the address computation based on which specific XcvrAddr it is. For backwards compatibility purposes we could define these in new APIs read_xcvr_eeprom/write_xcvr_eeprom.

Any thoughts on this approach?

Reply:
That does seem like a reasonable approach to me, but others may have differing opinions.

Reply:
As of today the linear address space assumption should be fine for all vendors and so I'll note this down as a future enhancement.